src/hammer-vlsi/test.py

```diff
@@ -616,6 +619,80 @@ def test_bumps(self) -> None:
+            out_dict["stackups"] = [tech_test.StackupTestHelper.create_test_stackup_dict(8)]
+            return out_dict
+         HammerToolTestHelpers.write_tech_json(tech_json_filename, add_stackup)
+         tech = self.get_tech(hammer_tech.HammerTechnology.load_from_dir("dummy28", tech_dir))
+         tech.cache_dir = tech_dir
+         tech.logger = HammerVLSILogging.context("")
+
+         test = DummyTool()
+         test.logger = HammerVLSILogging.context("")
+         test.run_dir = tempfile.mkdtemp()
+         test.technology = tech
+         database = hammer_config.HammerDatabase()
+         # Check bad mode string doesn't look at null dict
+         settings = """
+{
+     "vlsi.inputs.pin_mode": "auto"
+}
+"""
+         database.update_project([hammer_config.load_config_from_string(settings, is_yaml=False)])
+         test.set_database(database)
+
+         with HammerLoggingCaptureContext() as c:
+             my_pins = test.get_pin_assignments()
+         self.assertTrue(c.log_contains("Invalid pin_mode"))
+         assert len(my_pins) == 0, "Invalid pin_mode should assume empty pins"
+
+         settings = """
+ {
+     "technology.core.stackup": "StackupWith8Metals",
+     "vlsi.inputs.pin_mode": "generated",
+     "vlsi.inputs.pin.assignments": [
+                         {"pins": "*", "side": "top", "layers": ["M5", "M3"]},
+                         {"pins": "*", "side": "bottom", "layers": ["M5"]},
+                         {"pins": "*", "side": "left", "layers": ["M4"]},
+                         {"pins": "*", "side": "right", "layers": ["M2"]},
+                         {"pins": "bad_side", "side": "right", "layers": ["M3"]},
+                         {"pins": "tx_n tx_p", "macro": true},
+                         {"pins": "bad_tx_n", "macro": true, "layers": ["M7"]},
+                         {"pins": "tx_n tx_p", "macro": true, "side": "right", "layers": ["M7"]},
+                         {"pins": "tx_n tx_p", "macro": true, "side": "right"},
+                         {"pins": "*", "layers": ["M2"]},
+                         {"pins": "*", "side": "bottom"},
+                         {"pins": "no_layers"},
+                         {"pins": "wrong_side", "side": "upsidedown", "layers": ["M2"]}
+                     ]
+ }
+ """
+         database.update_project([hammer_config.load_config_from_string(settings, is_yaml=False)])
+         test.set_database(database)
+
+         with HammerLoggingCaptureContext() as c:
+             my_pins = test.get_pin_assignments()
+         self.assertTrue(c.log_contains("Pins bad_side assigned layers "))
+         self.assertTrue(c.log_contains("Pins bad_tx_n assigned as a macro pin with layers"))
+         self.assertTrue(c.log_contains("Pins no_layers assigned without layers"))
+         self.assertTrue(c.log_contains("Pins wrong_side have invalid side"))
+         assert my_pins is not None
+         # Only one of the assignments is invalid so the above 7 becomes 6
```

Comment:
Is there a test that captures what happens when all the values are good?

Reply:
Several of the pins above are "all good". It didn't seem helpful to break out the test cases since no problems are fatal.

Reply:
Yeah I'm thinking from a future proofing perspective where we add additional IR checks or make the current checks more strict it may be good to have a testcase for "good pin spec".

Reply:
Done